### PR TITLE
56 add db bill creation

### LINKED ISSUE #56: Add DB Bill Creation
Including:

- [x] Create Bill
- [x] Create Item
- [x] Add Item To Bill

### PR DESCRIPTION
**Pull request by f-odc**
Closes #56

### CODE REVIEW

File: presentation/dto/bill.go [hunk 1]
```diff
@@ -10,31 +10,26 @@ type BillInputDTO struct {
 	Owner uuid.UUID `json:"owner"`
 	Name  string    `json:"name"`
 	Date  time.Time `json:"date"`
-	Items []ItemDTO `json:"items"`
 	Group uuid.UUID `json:"group"`
```
Comment (GR0ZA) [inline presentation/dto/bill.go:13]: Can we rename the JSON property to groupId/groupID for convenience?

File: presentation/dto/bill.go [hunk 2]
```diff
@@ -10,31 +10,26 @@ type BillInputDTO struct {
 	Owner uuid.UUID `json:"owner"`
 	Name  string    `json:"name"`
 	Date  time.Time `json:"date"`
```
Comment (GR0ZA) [inline presentation/dto/bill.go:12]: What is the expected time format here?

**Review by GR0ZA — CHANGES_REQUESTED** [review]

File: presentation/dto/item.go [hunk 3]
```diff
@@ -5,20 +5,26 @@ import (
...
+type ItemInputDTO struct {
+	Name   string    `json:"name"`
+	Price  float64   `json:"price"`
+	BillID uuid.UUID `json:"billId"`
 }
 
-func ToItemModel(i ItemDTO) ItemModel {
-	return CreateItemModel(i.Name, i.Price, i.Contributors)
+type ItemOutputDTO struct {
+	Name  string  `json:"name"`
```
Comment (GR0ZA) [inline presentation/dto/item.go:15]: Would probably make sense to return the id of the item as well

File: storage/database/entity/bill.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"github.com/google/uuid"
+	. "split-the-bill-server/domain/model"
+	"time"
+)
+
+type Bill struct {
+	Base
+	Name    string `gorm:"not null"`
```
Comment (GR0ZA) [inline storage/database/entity/bill.go:11]: Don't you think we should store the ownerId as well?

**Comment by GR0ZA** [comment]
Creating a bill results in a response containing an empty bill DTO (can be fixed in another issue):

image

File: storage/database/entity/bill.go (hunk 4 above)
Comment (f-odc) [reply storage/database/entity/bill.go]: Yes good idea.

File: presentation/handler/bill_handler.go [hunk 5]
```diff
@@ -55,37 +55,62 @@ func (h BillHandler) GetByID(c *fiber.Ctx) error {
...
-	}
 
 	// parse bill from request body
+	var request BillInputDTO
 	err := c.BodyParser(&request)
 	if err != nil {
 		return core.Error(c, fiber.StatusBadRequest, fmt.Sprintf(ErrMsgBillParse, err))
 	}
 
+	// TODO: Who is responsible for validating the request? Handler or service?
```
[the remark this answers is not in the document]
Comment (f-odc) [reply presentation/handler/bill_handler.go:68]: I think we can delete it completely. Gorm can not create a bill entry if the foreignkey groupID is not correctly set. I suggest let this kind of validation be performed by gorm and add a more finer error handling to the storage.

Example:
if errors.Is(err, gorm.ErrForeignKeyViolated):
     return own storage errors

https://gorm.io/docs/error_handling.html

File: presentation/handler/bill_handler.go [hunk 6]
```diff
@@ -55,37 +55,62 @@ func (h BillHandler) GetByID(c *fiber.Ctx) error {
...
+func (h BillHandler) AddItem(c *fiber.Ctx) error {
+
+	// parse item from request body
+	var request ItemInputDTO
+	err := c.BodyParser(&request)
+	if err != nil {
+		return core.Error(c, fiber.StatusBadRequest, fmt.Sprintf(ErrMsgItemParse, err))
+	}
+
+	// TODO: Who is responsible for validating the request? Handler or service?
```
[the remark this answers is not in the document]
Comment (f-odc) [reply presentation/handler/bill_handler.go:102]: See above.

File: storage/database/db_storages/bill_storage.go [hunk 7]
```diff
@@ -17,11 +19,39 @@ func NewBillStorage(DB *database.Database) IBillStorage {
...
 func (b *BillStorage) GetByID(id uuid.UUID) (BillModel, error) {
-	//TODO implement me
-	panic("implement me")
+	var bill Bill
+	tx := b.DB.Limit(1).Preload("Items").Find(&bill, "id = ?", id)
+	if tx.RowsAffected == 0 {
+		return BillModel{}, storage.NoSuchBillError
+	}
+	// TODO: return general error
+	if tx.Error != nil {
```
[the remark this answers is not in the document]
Comment (f-odc) [reply storage/database/db_storages/bill_storage.go:40]: I would handle this problem in a separate issue in which we reconstruct the error handling of the storages.

File: presentation/handler/bill_handler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (GR0ZA) [reply presentation/handler/bill_handler.go]: Sounds good! If this kind of validation won't suffice for us we still can add some checks on service level later.

File: domain/model/bill.go [hunk 8]
```diff
@@ -10,15 +10,16 @@ type BillModel struct {
 	OwnerID uuid.UUID
 	Name    string
 	Date    time.Time
-	Items   []*ItemModel
+	Group   uuid.UUID
+	Items   []ItemModel
 }
 
-func CreateBill(ownerID uuid.UUID, name string, date time.Time, items []*ItemModel) BillModel {
+func CreateBill(ownerID uuid.UUID, name string, date time.Time, group uuid.UUID) BillModel {
```
[the remark this answers is not in the document]
Comment (f-odc) [reply domain/model/bill.go]: Let us keep it this way for now.

File: domain/model/group.go [hunk 9]
```diff
@@ -9,7 +9,7 @@ type GroupModel struct {
 	Owner   uuid.UUID
 	Name    string
 	Members []uuid.UUID
-	Bills   []BillModel
+	Bills   []uuid.UUID
```
[the remark this answers is not in the document]
Comment (f-odc) [reply domain/model/group.go:12]: We will work on this in #82

File: presentation/dto/group.go [hunk 10]
```diff
@@ -12,11 +12,11 @@ type GroupInputDTO struct {
...
-	Owner   UUID            `json:"owner"`
-	ID      UUID            `json:"id"`
-	Name    string          `json:"name"`
-	Members []UUID          `json:"members"`
-	Bills   []BillOutputDTO `json:"bills"`
+	Owner   UUID   `json:"owner"`
+	ID      UUID   `json:"id"`
+	Name    string `json:"name"`
+	Members []UUID `json:"members"`
+	Bills   []UUID `json:"bills"`
```
[the remark this answers is not in the document]
Comment (f-odc) [reply presentation/dto/group.go]: We will work on this in #82

File: presentation/dto/bill.go [hunk 11]
```diff
@@ -7,34 +7,35 @@ import (
...
+	Date  time.Time      `json:"date"`
+	Group uuid.UUID      `json:"groupID"`
+	Items []ItemInputDTO `json:"items"`
 }
 
 type BillOutputDTO struct {
-	Name  string    `json:"name"`
-	Date  time.Time `json:"date"`
-	Items []ItemDTO `json:"items"`
+	Name  string          `json:"name"`
```
Comment (GR0ZA) [inline presentation/dto/bill.go:18]: Add Bill ID to BillOutputDTO

File: storage/database/entity/bill.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"github.com/google/uuid"
+	. "split-the-bill-server/domain/model"
+	"time"
+)
+
+type Bill struct {
+	Base
+	OwnerUID uuid.UUID `gorm:"type:uuid"`
```
Comment (GR0ZA) [inline storage/database/entity/bill.go:11]: Rename to OwnerID for consistency.

File: storage/database/entity/group.go [hunk 13]
```diff
@@ -11,6 +11,7 @@ type Group struct {
 	OwnerUID uuid.UUID `gorm:"type:uuid"`
 	Owner    User      `gorm:"foreignKey:OwnerUID"`
 	Members  []*User   `gorm:"many2many:group_members"`
+	Bills    []Bill    `gorm:"foreignKey:GroupID"` // has many bills
```
Comment (GR0ZA) [inline storage/database/entity/group.go:14]: I like these comments, they make it more descriptive what is happening here :)

**Review by GR0ZA — CHANGES_REQUESTED** [review]

File: storage/database/entity/group.go (hunk 13 above)
Comment (f-odc) [reply storage/database/entity/group.go:14]: yippie!

**Pull request merged**